docs/okeps/okep-5259-no-overlay.md

```diff
@@ -0,0 +1,595 @@
+  # 'default' is the reserved name to represent the cluster default network
+  name: default
+spec:
+  # use an empty selector to select all namespaces
+  namespaceSelector: {}
+  network:
+    topology: Layer3
+    layer3:
+      role: Primary
+      mtu: 1500
+      subnets:
+      - cidr: 10.244.0.0/16
+        hostSubnet: 24
+      encapsulation: None
+```
+
+### Implementation Details
+
+No-overlay mode is only available when OVNK interconnect is enabled, since it
+relays on the route advertisement feature, which only works with interconnect
+mode.
+
+A network that is configured to use no-overlay mode needs a slightly modified
+OVN topology. It is useful to remember that in today's interconnect architecture
+a network is able to span all cluster nodes thanks to the cross-node logical
+links that OVN establishes between the transit switch and the ovn-cluster-router
+instances on each node. That is the logical representation of the OVN network
+and in practice these links are implemented through Geneve tunnels between every
+node in the cluster. For a network to skip the transport step and route pod
+traffic directly over the underlay network, we need to change the logical
+network topology of this network: its ovn-cluster-router won't connect to the
+transit switch anymore and all ease-west traffic will be naturally routed in the
+same path as north-south traffic through the provider network.
+
+In Shared Gateway Mode, at the OVN ClusterRouter, the pod to remote pod egress
+traffic will be forwarded to the GatewayRouter via the join switch. Different
+from the current route advertisement behavior, the routes to remote node subnets
+will be imported to the Gateway Router. Once the traffic reach the
+GatewayRouter, it will be forwarded the nexthop node according to routes learnt
+from BGP.
+
+In Local Gateway Mode, the pod to remote pod traffic egress traffic will enter
+the host VRF via the ovn-k8s-mpX interface. Then it will be forwarded to remote
+nodes according to the host routes of the VRF.
+
+The following diagram shows the OVN topology of a cluster with three networks:
+
+* a CUDN (in green) that uses Geneve transport (default behavior)
+* a default network (in gray) that enables no-overlay mode
+* a CUDN (in blue) that enables no-overlay mode
+
+![No-overlay topology](../images/no_overlay_topology.jpg)
+
+#### No SNAT in no-overlay mode for egress traffic
+
+The BGP feature in ovn-kubernetes only disables SNAT for traffic that is
+destined to external networks, not for traffic that is destined to other pods in
+the cluster.
+
+In no-overlay mode, we want to fully leverage BGP and avoid the SNAT step on the
+gateway router for intra-cluster traffic as well. Pod-to-pod traffic within the
+same network will hit the BGP-learned routes to reach its destination node,
+without being encapsulated in a Geneve tunnel and without being SNATed to the
+node IP.
+
+The only exception is for pod to other host traffic, to ensure nodePort services
```

Comment:
Uhm here I don't follow. Are we keeping the SNAT so that we can allow traffic to nodeports on different nodes, or are we keeping the SNAT so that users won't expose pod IPs to the outside?

Reply:
We keep the SNAT only for pod-to-remote_host traffic.